webapp/src/components/backstage/playbook_runs/playbook_run/timeline_utils.ts

```diff
@@ -0,0 +1,151 @@
+    const displayPreference = useSelector<GlobalState, string | undefined>(getTeammateNameDisplaySetting) || 'username';
+    const [allEvents, setAllEvents] = useState<TimelineEvent[]>([]);
+    const [filteredEvents, setFilteredEvents] = useState<TimelineEvent[]>([]);
+    const eventsFilter = useSelector<GlobalState, TimelineEventsFilter>((state) => eventsFilterForPlaybookRun(state, playbookRun.id));
+    const getStateFn = useStore().getState;
+    const getUserFn = (userId: string) => getUserAction(userId)(dispatch as DispatchFunc, getStateFn);
+    const selectUser = useSelector<GlobalState, IdToUserFn>((state) => (userId: string) => getUser(state, userId));
+
+    useEffect(() => {
+        setFilteredEvents(allEvents.filter((e) => showEvent(e.event_type, eventsFilter)));
+    }, [eventsFilter, allEvents]);
+
+    useEffect(() => {
+        const {
+            status_posts: statuses,
+            timeline_events: events,
+        } = playbookRun;
+
+        const statusDeleteAtByPostId = statuses.reduce<{[id: string]: number}>((map, post) => {
+            if (post.delete_at !== 0) {
+                map[post.id] = post.delete_at;
+            }
+
+            return map;
+        }, {});
+
+        Promise.all(events.map(async (event) => {
+            let user = selectUser(event.subject_user_id) as UserProfile | undefined;
+
+            if (!user) {
+                const ret = await getUserFn(event.subject_user_id) as { data?: UserProfile, error?: any };
+                if (!ret.data) {
+                    return null;
+                }
+                user = ret.data;
+            }
+            return {
+                ...event,
+                status_delete_at: statusDeleteAtByPostId[event.post_id] ?? 0,
+                subject_display_name: displayUsername(user, displayPreference),
+            } as TimelineEvent;
+        })).then((eventArray) => {
+            eventArray.reverse();
+            setAllEvents(eventArray.filter((e) => e) as TimelineEvent[]);
+        });
+    }, [playbookRun.timeline_events, displayPreference, playbookRun.status_posts]);
+
+    return [filteredEvents];
+};
+
+const showEvent = (eventType: string, filter: TimelineEventsFilter) => {
+    if (filter.all) {
+        return true;
+    }
+    const filterRecord = filter as unknown as Record<string, boolean>;
+    return filterRecord[eventType] ||
+        (eventType === TimelineEventType.RunCreated && filterRecord[TimelineEventType.StatusUpdated]) ||
+        (eventType === TimelineEventType.RunFinished && filterRecord[TimelineEventType.StatusUpdated]) ||
+        (eventType === TimelineEventType.RunRestored && filterRecord[TimelineEventType.StatusUpdated]);
+};
+
+export const useFilter = (playbookRun: PlaybookRun) => {
+    const {formatMessage} = useIntl();
+    const dispatch = useDispatch();
+    const eventsFilter = useSelector<GlobalState, TimelineEventsFilter>((state) => eventsFilterForPlaybookRun(state, playbookRun.id));
```

Comment:
I'm the most annoying reviewer, sorry
```suggestion
    const eventsFilter = useSelector((state: GlobalState) => eventsFilterForPlaybookRun(state, playbookRun.id));
```